src/caffe/layers/detectnet_transform_layer.cpp

```diff
@@ -0,0 +1,929 @@
+    phase_(param.phase()),
+    rng_(new Caffe::RNG(caffe_rng_rand()))
+{
+}
+
+
+template <typename Dtype>
+void DetectNetTransformationLayer<Dtype>::LayerSetUp(
+    const vector<Blob<Dtype>*>& bottom,
+    const vector<Blob<Dtype>*>& top
+) {
+  // retrieve mean image or values:
+  if (t_param_.has_mean_file()) {
+    size_t image_x = bottom[0]->width();
+    size_t image_y = bottom[0]->height();
+
+    retrieveMeanImage(Size(image_x, image_y));
+
+  } else if (t_param_.mean_value_size() != 0) {
+    retrieveMeanChannels();
+  }
+}
+
+
+template<typename Dtype>
+void DetectNetTransformationLayer<Dtype>::retrieveMeanImage(Size dimensions) {
+  CHECK(t_param_.has_mean_file());
+
+  const string& mean_file = t_param_.mean_file();
+  BlobProto blob_proto;
+  Blob<Dtype> data_mean;
+
+  ReadProtoFromBinaryFileOrDie(mean_file, &blob_proto);
+  data_mean.FromProto(blob_proto);
+  data_mean_ = blobToMats(data_mean).at(0);
+
+  // resize, if dimensions were defined:
+  if (dimensions.area() > 0) {
+    resize(data_mean_, data_mean_, dimensions, cv::INTER_CUBIC);
+  }
+  // scale from 0..255 to 0..1:
+  data_mean_ /= Dtype(UINT8_MAX);
+}
+
+
+template<typename Dtype>
+void DetectNetTransformationLayer<Dtype>::retrieveMeanChannels() {
+  switch(t_param_.mean_value_size()) {
+    case 1:
+      mean_values_.fill(t_param_.mean_value(0) / Dtype(UINT8_MAX));
+      break;
+    case 3:
+      for (size_t iChannel = 0; iChannel != 3; ++iChannel) {
+        mean_values_[iChannel] =
+            t_param_.mean_value(iChannel) / Dtype(UINT8_MAX);
+      }
+      break;
+    case 0:
+    default:
+      break;
+  }
+}
+
+
+template<typename Dtype>
+void DetectNetTransformationLayer<Dtype>::Reshape(
+    const vector<Blob<Dtype>*>& bottom,
+    const vector<Blob<Dtype>*>& top
+) {
+  // accept only three channel images:
+  CHECK_EQ(bottom[0]->channels(), 3);
+  // accept only equal numbers of labels and images:
+  CHECK_EQ(bottom[0]->num(), bottom[1]->num());
+  // resize mean image if it exists:
+  if (t_param_.has_mean_file()) {
+    size_t image_x = bottom[0]->width();
+    size_t image_y = bottom[0]->height();
+    retrieveMeanImage(Size(image_x, image_y));
+  }
+  // resize image output layer: (never changes /wrt input blobs)
+  top[0]->Reshape(
+      bottom[0]->num(),
+      bottom[0]->channels(),
+      g_param_.image_size_y(),
+      g_param_.image_size_x()
+  );
+  // resize tensor output layer: (never changes /wrt input blobs)
+  Vec3i tensorDimensions = coverage_->dimensions();
+  top[1]->Reshape(
+      bottom[1]->num(),
+      tensorDimensions(0),
+      tensorDimensions(1),
+      tensorDimensions(2)
+  );
+}
+
+
+template<typename Dtype>
+typename DetectNetTransformationLayer<Dtype>::Mat3v
+DetectNetTransformationLayer<Dtype>::dataToMat(
+    const Dtype* _data,
+    Size dimensions
+) const {
+  // CODE FROM https://github.com/BVLC/caffe/blob/master/examples/cpp_classification/classification.cpp commit f3a933a
```

Comment:
Don't make links to GitHub with the branch name in the URL because the code will change.

Instead, link to a specific line at a specific like this:

https://github.com/BVLC/caffe/blob/4874c01487/examples/cpp_classification/classification.cpp#L1-L4